storage/testonly/fake_node_reader.go

```diff
@@ -0,0 +1,196 @@
+// package as this creates import cycles.
+
+// NodeMapping is a struct we use because we can't use NodeIDs as map keys. Callers pass this
+// and FakeNodeReader internally manages derived keys.
+type NodeMapping struct {
+	NodeID storage.NodeID
+	Node   storage.Node
+}
+
+// FakeNodeReader is an implementation of storage.NodeReader that's preloaded with a set of
+// NodeID -> Node mappings and will return only those. Requesting any other nodes results in
+// an error. For use in tests only, does not implement any other storage APIs.
+type FakeNodeReader struct {
+	treeSize     int64
+	treeRevision int64
+	nodeMap      map[string]storage.Node
+}
+
+// NewFakeNodeReader creates and returns a FakeNodeReader with the supplied nodeID -> Node
+// mappings assuming that all the nodes are at a specified tree revision. All the nodeIDs
+// must be distinct.
+func NewFakeNodeReader(mappings []NodeMapping, treeSize, treeRevision int64) *FakeNodeReader {
+	nodeMap := make(map[string]storage.Node)
+
+	for _, mapping := range mappings {
+		_, ok := nodeMap[mapping.NodeID.String()]
+
+		if ok {
+			// Duplicate mapping - bail out
+			return nil
+		}
+
+		nodeMap[mapping.NodeID.String()] = mapping.Node
+	}
+
+	return &FakeNodeReader{nodeMap: nodeMap, treeSize: treeSize, treeRevision: treeRevision}
+}
+
+// GetTreeRevisionAtSize implements the corresponding NodeReader API
+func (f FakeNodeReader) GetTreeRevisionAtSize(treeSize int64) (int64, error) {
+	if f.treeSize != treeSize {
+		return int64(0), fmt.Errorf("GetTreeRevisionAtSize() got treeSize:%d, want: %d", treeSize, f.treeSize)
+	}
+
+	return f.treeRevision, nil
+}
+
+// GetMerkleNodes implements the corresponding NodeReader API
+func (f FakeNodeReader) GetMerkleNodes(treeRevision int64, NodeIDs []storage.NodeID) ([]storage.Node, error) {
+	if f.treeRevision > treeRevision {
+		return []storage.Node{}, fmt.Errorf("GetMerkleNodes() got treeRevision:%d, want up to: %d", treeRevision, f.treeRevision)
+	}
+
+	nodes := make([]storage.Node, 0, len(NodeIDs))
+	for _, nodeID := range NodeIDs {
+		node, ok := f.nodeMap[nodeID.String()]
+
+		if !ok {
+			return []storage.Node{}, fmt.Errorf("GetMerkleNodes() unknown node ID: %v", nodeID)
+		}
+
+		nodes = append(nodes, node)
+	}
+
+	return nodes, nil
+}
+
+func (f FakeNodeReader) hasID(nodeID storage.NodeID) bool {
+	_, ok := f.nodeMap[nodeID.String()]
+	return ok
+}
+
+// MultiFakeNodeReader can provide nodes at multiple revisions. It delegates to a number of
+// FakeNodeReaders, each set up to handle one revision.
+type MultiFakeNodeReader struct {
+	readers []FakeNodeReader
+}
+
+// LeafBatch describes a set of leaves to be loaded into a MultiFakeNodeReader via a compact
+// merkle tree. As each batch is added to the tree a set of node updates are collected
+// and recorded in a FakeNodeReader for that revision.
+type LeafBatch struct {
+	TreeRevision int64
+	Leaves       []string
+	ExpectedRoot string
```

Comment:
nit: Mention ExpectedRoot's encoding, or type as []byte and use bytes.Compare() instead?